### DISCUSSION TITLE
How does Next.js 14 Partial Prerendering Work? No AJAX Requests?

### DISCUSSION THREAD

**Discussion by maximilianschmitt** [vercel/next.js#57539] [Help]
In the [announcement post](https://nextjs.org/blog/next-14) for Next.js 14, there is a section about a new feature called [Partial Prerendering](https://nextjs.org/blog/next-14#partial-prerendering-preview). It says:

> Since `<ShoppingCart />` reads from cookies to look at the user session, this component is then **streamed in as part of the same HTTP request as the static shell. There are no extra network roundtrips needed.**

How does this work? How can you stream content into holes of an HTML document without making AJAX requests? I always thought that HTML streaming responses have to serve the markup in order.

**Comment by icyJoseph** [comment]
Look at this: https://cursed-next-patterns.t3.gg/generators

Open the network, and you can see how that, otherwise fringe demo, does it.

**Reply by maximilianschmitt** [reply]
Interesting! It looks like it basically just streams in chunks of JS that update the site piece-by-piece. If that's how the partial prerendering works, then it won't be completely SSR and it won't work without JavaScript.

**Reply by icyJoseph** [reply]
We'll see, I don't have many details nor have I tested it, but I can speculate:

- a `noscript` tag could be used to tell the server that the page can't do the tecnique
- If so, then fallback on what bots/seo would see, which would make the server await the entire tree and flush it out

There's a whole bunch on the HTTP 2 and streaming, that eventually, we'll have to learn :/ I am no expert on that area just yet

**Reply by jherr** [reply]
I still don't get it. I see what's happening in the HTML, it's getting the non-suspense chunks first and then streaming in the rest by adding more script blocks and HTML, totally makes sense. The question is how is the CDN involved? The docs say this the "Vercel Edge Network". But maybe the reason it's so fast is that it's an edge function that first gets the static chunk and sends that back while starting the processing of the suspenses?

**Comment by Daniel3711997** [comment]
what is the difference between partial rendering and the current streamming option of app router ? 🧐

**Reply by icyJoseph** [reply]
Partial pre-rendering can be done at build time.

**Reply by Daniel3711997** [reply]
thank you
